src/schedule/ll_schedule.c

```diff
@@ -72,6 +72,14 @@ static bool schedule_ll_is_pending(struct ll_schedule_data *sch)
 		list_for_item(tlist, &sch->tasks) {
 			task = container_of(tlist, struct task, list);
 
+			/*
+			 * only tasks queued or waiting for reschedule are
+			 * pending for scheduling
+			 */
+			if (task->state != SOF_TASK_STATE_QUEUED &&
+			    task->state != SOF_TASK_STATE_RESCHEDULE)
+				continue;
```

Comment:
Could you explain what this is fixing? IIUC currently all tasks on the list are queued, should run. If a task completes or is cancelled it's immediately removed from the list.

Reply:
This doesn't fix an observed/existed issue, as schedule_ll_is_pending() is a helper to update tasks to be pending here, it is only to make the logic more clear and avoid misusing of it. E.g. if a task is in running while someone call schedule_ll_is_pending(), the state of this running task could be updated to _PENDING by mistake.

Yes, this will not happen with today's usage, we call schedule_ll_is_pending() in a scheduler's interrupt handling, as the arrival of a specific DSP core means all tasks running on it has been finished/rescheduled, and tasks belong to other cores are *not* in this sch->tasks list. The naming of "schedule_ll_is_pending" is somewhat confusion, it doesn't check if tasks are pending, it actually move tasks (should be in _PENDING) to be _PENDING.